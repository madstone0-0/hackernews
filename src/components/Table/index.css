.table {
    margin: 20px 0;
}

.table-header {
    display: flex;
    line-height: 24px;
    font-size: 16px;
    padding: 0 10px;
    justify-content: space-between;
}

.title-header {
    width: 40%;
}

.author-header {
    width: 30%
}

.comments-header, .points-header, ,archive-header {
    width: 10%
}

.table-empty {
    margin: 200px;
    text-align: center;
    font-size: 16px;
}

.table-row {
    display: flex;
    line-height: 24px;
    white-space: nowrap;
    margin: 10px 0;
    padding: 10px;
    background: #ffffff;
    border: 1px solid #e3e3e3;
}

.table-header > span {
    overflow: hidden;
    text-overflow: ellipsis;
    padding: 0 5px;
}

.table-row > span {
    overflow: hidden;
    text-overflow: ellipsis;
    padding: 0 5px;
}

.button-inline {
    border-width: 0;
    background: transparent;
    color: inherit;
    text-align: inherit;
    -webkit-font-smoothing: inherit;
    padding: 0;
    font-size: inherit;
    cursor: pointer;
}

.button-active {
    border-radius: 0;
    border-bottom: 1px solid #82b1ff;
}

/* Dark Table styles */
.dark .page .table {
    margin: 20px 0;
}

.dark .page .table-row {
    display: flex;
    line-height: 24px;
    white-space: nowrap;
    margin: 10px 0;
    padding: 10px;
    background: #757575;
    border: 1px solid #424242;
}

.dark .page .button-active {
    border-radius: 0;
    border-bottom: 1px solid #ff1744;
}

@media only screen and (max-width: 768px) {
    .table-header {
        font-size: 12px;
        padding: 0 10px;
    }

    .table-row {
        font-size: 12px;
    }

    .title-header {
        width: 20%;
    }
    
    .author-header {
        width: 15%
    }
    
    .comments-header, .points-header, ,archive-header {
        width: 5%
    }
}