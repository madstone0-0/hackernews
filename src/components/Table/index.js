/* eslint-disable no-unused-vars,no-undef */
import React, { Component } from "react";
import Button from "../Button";
import "./index.css";
import PropTypes from "prop-types";
import classNames from "classnames";
import { sortBy } from "lodash";

const SORTS = {
    NONE: list => list,
    TITLE: list => sortBy(list, "title"),
    AUTHOR: list => sortBy(list, "author"),
    COMMENTS: list => sortBy(list, "num_comments").reverse(),
    POINTS: list => sortBy(list, "points").reverse(),
};

/**
 * Sort component
 * (not sure if should be refactored or not)
 */
const Sort = ({ sortKey, onSort, children, activeSortKey }) => {
    const sortClass = classNames("button-inline", {
        "button-active": sortKey === activeSortKey,
    });

    return (
        <Button onClick={() => onSort(sortKey)} className={sortClass}>
            {children}
        </Button>
    );
};

Sort.propTypes = {
    sortKey: PropTypes.string.isRequired,
    onSort: PropTypes.func.isRequired,
    children: PropTypes.string,
    activeSortKey: PropTypes.string,
};

const largeColumn = {
    width: "40%",
};
const midColumn = {
    width: "30%",
};
const smallColumn = {
    width: "10%",
};

/**
 * Table component
 */
class Table extends Component {
    constructor(props) {
        super(props);

        this.state = {
            sortKey: "NONE",
            isSortReverse: false,
        };

        this.onSort = this.onSort.bind(this);
    }

    onSort(sortKey) {
        const isSortReverse =
            this.state.sortKey === sortKey && !this.state.isSortReverse;
        this.setState({ sortKey, isSortReverse });
    }

    render() {
        const { list, onDismiss } = this.props;

        const { sortKey, isSortReverse } = this.state;

        const sortedList = SORTS[sortKey](list);
        const reverseSortedList = isSortReverse
            ? sortedList.reverse()
            : sortedList;

        return (
            <div className="table">
                <div className="table-header">
                    <span className="title-header">
                        <Sort
                            sortKey={"TITLE"}
                            onSort={this.onSort}
                            activeSortKey={sortKey}
                        >
                            Title
                        </Sort>
                    </span>
                    <span className="author-header">
                        <Sort
                            sortKey={"AUTHOR"}
                            onSort={this.onSort}
                            activeSortKey={sortKey}
                        >
                            Author
                        </Sort>
                    </span>
                    <span className="comments-header">
                        <Sort
                            sortKey={"COMMENTS"}
                            onSort={this.onSort}
                            activeSortKey={sortKey}
                        >
                            Comments
                        </Sort>
                    </span>
                    <span className="points-header">
                        <Sort
                            sortKey={"POINTS"}
                            onSort={this.onSort}
                            activeSortKey={sortKey}
                        >
                            Points
                        </Sort>
                    </span>
                    <span className="archive-header">Archive</span>
                </div>
                {reverseSortedList.map(item => {
                    return (
                        <div
                            key={item.objectID}
                            className="table-row col-xs-12"
                        >
                            <span style={largeColumn}>
                                <a
                                    target="_blank"
                                    rel="noopener noreferrer"
                                    href={item.url}
                                >
                                    {item.title}
                                </a>
                            </span>
                            <span style={midColumn}>{item.author}</span>
                            <span style={smallColumn}>{item.num_comments}</span>
                            <span style={smallColumn}>{item.points}</span>
                            <span style={smallColumn}>
                                <Button
                                    onClick={() => onDismiss(item.objectID)}
                                    className="button-inline"
                                >
                                    Dismiss
                                </Button>
                            </span>
                        </div>
                    );
                })}
            </div>
        );
    }
}

Table.propTypes = {
    list: PropTypes.arrayOf(
        PropTypes.shape({
            objectID: PropTypes.string.isRequired,
            author: PropTypes.string,
            url: PropTypes.string,
            num_comments: PropTypes.number,
            points: PropTypes.number,
        }),
    ).isRequired,
    onDismiss: PropTypes.func.isRequired,
};

export default Table;
