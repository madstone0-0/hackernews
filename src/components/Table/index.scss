.table {
    margin: 20px 0;

    .table-header {
        display: flex;
        line-height: 24px;
        font-size: 16px;
        padding: 0 10px;
        justify-content: space-between;
    }

    .title-header {
        width: 38%;
    }

    .author-header {
        width: 30%;
        padding-left: 30px;
    }

    .comments-header, .points-header, .archive-header {
        width: 10%;
    }

    .comments-header {
        margin-left: 15px;
    }

    // .archive-header {
    //     margin-right: 69px;
    // }

    .table-empty {
        margin: 200px;
        text-align: center;
        font-size: 16px;
    }

    .table-row {
        display: flex;
        line-height: 24px;
        white-space: nowrap;
        margin: 10px 0;
        padding: 10px;
        background: #ffffff;
        border: 1px solid #e3e3e3;
    }

    .table-header > span, .table-row > span {
        overflow: hidden;
        text-overflow: ellipsis;
        padding: 0 5px;
    }

    .button-inline {
        border-width: 0;
        background: transparent;
        color: inherit;
        text-align: inherit;
        -webkit-font-smoothing: inherit;
        padding: 0;
        font-size: inherit;
        cursor: pointer;
    }

    .button-active {
        border-radius: 0;
        border-bottom: 1px solid #82b1ff;
    }

    @media only screen and (max-width: 768px) {
        .table-header {
            font-size: 14px;
            padding: 0 10px;
        }

        .table-row {
            font-size: 12px;
        }

        .title-header {
            width: 20%;
        }

        .author-header {
            width: 15%;
        }

        .comments-header, .points-header, .archive-header {
            width: 5%;
        }
    }

}

/* Dark Table styles */

.dark {
    .page {
        .table {
            margin: 20px 0;

            .table-row {
                display: flex;
                line-height: 24px;
                white-space: nowrap;
                margin: 10px 0;
                padding: 10px;
                background: #757575;
                border: 1px solid #424242;
            }

            .button-active {
                border-radius: 0;
                border-bottom: 1px solid #ff1744;
            }
        }
    }
}
