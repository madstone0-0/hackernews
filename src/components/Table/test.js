/* eslint-disable no-unused-vars,no-undef */
import ReactDOM from "react-dom";
import { shallow } from "enzyme/build";
import renderer from "react-test-renderer";
import React from "react";
import Table from "./index";
import Adadpter from "enzyme-adapter-react-16";
import * as Enzyme from "enzyme";

const props = {
    list: [
        { title: "1", author: "1", num_comments: 1, points: 2, objectID: "y" },
        { title: "2", author: "2", num_comments: 1, points: 2, objectID: "z" },
    ],
    sortKey: "TITLE",
    isSortReverse: false,
};

describe("Table", () => {

    it("renders", () => {
        const div = document.createElement("div");
        ReactDOM.render(<Table {...props}/>, div);
    });

    Enzyme.configure({ adapter: new Adadpter() });

    it("shows two items in a list", () => {
        const element = shallow(
            <Table {...props} />,
        );

        expect(element.find(".table-row").length).toBe(2);
    });
});

test("snapshots", () => {
    const component = renderer.create(
        <Table {...props}/>,
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});