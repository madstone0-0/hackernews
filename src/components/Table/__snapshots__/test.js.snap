// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshots 1`] = `
<div
  className="table"
>
  <div
    className="table-header"
  >
    <span
      style={
        Object {
          "width": "40%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
         
        Title
         
      </button>
    </span>
    <span
      style={
        Object {
          "width": "30%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
         
        Author
         
      </button>
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
         
        Comments
         
      </button>
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
         
        Points
         
      </button>
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      Archive
    </span>
  </div>
  <div
    className="table-row row"
  >
    <span
      className="col-md-12"
      style={
        Object {
          "width": "40%",
        }
      }
    >
      <a
        className="col-md-12"
      >
        1
      </a>
    </span>
    <span
      className="col-md-12"
      style={
        Object {
          "width": "30%",
        }
      }
    >
      1
    </span>
    <span
      className="col-md-12"
      style={
        Object {
          "width": "10%",
        }
      }
    >
      1
       comments
    </span>
    <span
      className="col-md-12"
      style={
        Object {
          "width": "10%",
        }
      }
    >
      2
       points
    </span>
    <span
      className="col-md-12"
      style={
        Object {
          "width": "10%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        Dismiss
      </button>
    </span>
  </div>
  <div
    className="table-row row"
  >
    <span
      className="col-md-12"
      style={
        Object {
          "width": "40%",
        }
      }
    >
      <a
        className="col-md-12"
      >
        2
      </a>
    </span>
    <span
      className="col-md-12"
      style={
        Object {
          "width": "30%",
        }
      }
    >
      2
    </span>
    <span
      className="col-md-12"
      style={
        Object {
          "width": "10%",
        }
      }
    >
      1
       comments
    </span>
    <span
      className="col-md-12"
      style={
        Object {
          "width": "10%",
        }
      }
    >
      2
       points
    </span>
    <span
      className="col-md-12"
      style={
        Object {
          "width": "10%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        Dismiss
      </button>
    </span>
  </div>
</div>
`;
